Switch to using Enum to track Entry state

gazpachoking opened the pull request:
### Motivation for changes:
While adding some type hints, I realized there was already an `EntryState` enum, and thought maybe we should start actually using it to store entry state.

There was some hacky stuff in digest plugin regarding entry state, I didn't even try to fix that in this pull, just updated it to still work.

liiight commented on flexget/plugins/operate/digest.py:
```diff
@@ -96,10 +96,11 @@ def on_task_learn(self, task, config):
             for entry in task.all_entries:
-                if entry.state not in config['state']:
+                state = str(entry.state)
+                if state not in config['state']:
```
hmm, maybe a custom json schema format that returns `EntryState`?

gazpachoking replied:
Not sure what you mean?

gazpachoking commented:
> This seems very nice. I wonder if there could be unforeseen effects to this via config. Maybe people have checks that rely on entry states being strings? does that make sense?

I added a custom `__str__` so that when used in a string context (jinja template, etc.) it should come out the same.

liiight commented:
> I added a custom __str__ so that when used in a string context (jinja template, etc.) it should come out the same.

Maybe add a test for this?

gazpachoking commented:
> > I added a custom **str** so that when used in a string context (jinja template, etc.) it should come out the same.
> 
> Maybe add a test for this?

Hmm, actually, it could be an issue inside a jinja expression, like in the `if` plugin. I'll add some backwards compatibility and a test for that.

gazpachoking commented:
> Hmm, actually, it could be an issue inside a jinja expression, like in the `if` plugin. I'll add some backwards compatibility and a test for that.

Strike that, we can't access entry state from the if plugin (although, perhaps that should be added.) Are there any other places outside of string formatting this could be an issue anyone can think of?

gazpachoking commented:
I added in `__eq__` and `__hash__` which should make the enum states pretty compatible with the old plain string states, even though I'm not sure there's anywhere in the config where that could matter. I'm thinking we should allow accessing the state from jinja though, probably in a separate pull. The compatibility stuff I just added allows `entry.state == 'accepted'` and `entry.state in ['accepted', 'rejected']` type stuff again.

gazpachoking commented:
Held off on any explicit tests, since we aren't actually relying on any special behavior of the entry state beyond what normal tests verify (that I know of.) If we allow it in jinja later, we can add some tests to verify that behavior works as expected.
